poncho/src/poncho/package_serverize.py

```diff
@@ -172,18 +167,29 @@ def pack_library_code(path, envpath):
 # Python's exec or Jupyter Notebooks won't work here.
 # @param functions  A list of functions to generate the hash value from.
 # @return           a string of hex characters resulted from hashing the contents and names of functions.
-def generate_functions_hash(functions: list) -> str:
+def generate_functions_hash(functions: list, import_modules=None) -> str:
+    import re
+    import sys
+
     source_code = ""
+    if import_modules:
+        source_code += " ".join(["import " + module.__name__ + "\n" for module in import_modules])
+
     for fnc in functions:
         try:
-            source_code += fnc.__name__
-            source_code += inspect.getsource(fnc)
+            # get function source code, remove single-line comments, blank lines and excessive newlines
```

Comment:
Why are you removing the comments, etc?  They don't hurt the hash computation below, and instead have the potential of introducing bugs as it is easy to get regular expressions wrong. For example, your expressions don't take into account that # may be inside a string.  I recommend against this change.

Reply:
That's a fair point. I thought sometimes the users may somehow enter a few random blanks or newlines, but they might don't want to re-install the environment as that takes several minutes but has the same effect with the previous one. You are right that the regular expressions are easy to introduce bugs, I will go ahead and remove it!

Reply:
If they modify the function in any way, then that's ok. It is better to be safe and recompute the hash rather than try to interpret if two functions are the same.

Reply:
Copy that!